Flash loan vault

hashedMae opened the pull request:
-yield bearing flash loan vault for a predetermined token
-removal of exchange rate and now uses proportional ownership
-adding of flash loans compliant to EIP 3156
-vault made compliant with EIP 4626
-update of testing
-still need to add testing for flash loans

hashedMae committed:
forge install: yield-utils-v2

hashedMae committed:
forge install: forge-std

hashedMae committed:
starting on assignment 3

hashedMae committed:
fuzz testing

hashedMae committed:
bugs

hashedMae committed:
corrections

hashedMae committed:
contract and testing

hashedMae committed:
made changes based on comments

hashedMae committed:
fixing math library

hashedMae committed:
-removal of exchange rate
-changed to proportional ownership
-adding of flash loans conforming to 3156
-conforms to 4246

hashedMae committed:
tests and changes based on commetns from Issue  #7

hashedMae committed:
logic fixing

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    /// @param shares amount of shares that are being burned
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    IERC20 public immutable iRings;
```
You could just call this `asset` and remove the function below. For ERC compliance purposes, any contract type is the same as `address`.

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    /// @param receiver address of user that would receive the tokens
+    /// @return maxAssets max amount of assets that can be deposited
+    function maxDeposit(address receiver) external pure returns(uint256) {
+        if(receiver == address(0x0)){
```
Why can't we deposit with `address(0)` as a `receiver`?

hashedMae replied:
To prevent tokens being burned on deposit.
This is now reflected in the deposit functions

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    /// @param receiver address of user that would receive the shares
+    /// @return maxShares max amount of shares a user can receive
+    function maxMint(address receiver) external pure returns(uint256) {
+        if(receiver == address(0x0)){
```
Same as above, why?

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+            "IERC3156: Callback failed"
+        );
+        uint256 _allowance = iRings.allowance(address(receiver), address(this));
+        require(
```
`transferFrom` already checks that the allowance is at least `amount + fee`, right? You don't need to check again.

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+        return true;
+    }
+
+    function init(uint256 assets, address receiver) external returns(uint256) {
```
You probably need a check that `init` can only be called if `totalSupply == 0`. Either that or init in the constructor.

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    /// @param assets Number of asset tokens to deposit
+    /// @param receiver Address that vault tokens will be minted to
+    /// @return shares Number of vault shares to return to the user
+    function deposit(uint256 assets, address receiver) external  returns(uint256) {
```
```suggestion
    function deposit(uint256 assets, address receiver) external  returns(uint256 shares) {
```

I find the syntax cleaner if the return variables have names and we skip the `return` statement

hashedMae replied:
I think I'd adopted this pattern since writing the first version of the vault. But it didn't stand out to me on the second version enough to change it.

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    /// @return shares Number of vault shares minted to the user based on the exchange rate
+    function _convertToShares(uint256 assets) internal view returns(uint256 shares) {
+        uint256 _reserves = iRings.balanceOf(address(this));
+        shares = WMul.wmul(WDiv.wdiv(assets, _reserves), _totalSupply);
```
This is equivalent to:
`(((assets * 1e18) / _reserves) * _totalSupply) / 1e18`

and therefore:
`assets * totalSupply / _reserves`

There is a chance for a phantom overflow (`assets * totalSupply` overflows, even if the result after dividing by `_reserves` fits), look for `mulDiv` implementations that would solve the issue.

alcueca commented on src/WRings.sol:
```diff
@@ -0,0 +1,268 @@
+    function _convertToAssets(uint256 shares) internal view returns(uint256 assets) {
+       uint256 _reserves = iRings.balanceOf(address(this));
+       uint256 _portion = WDiv.wdiv(shares, _totalSupply);
+       assets = WMul.wmul(_portion, _reserves);
```
Same as before:
`assets = (shares * _reserves) / _totalSupply`

alcueca reviewed (changes requested):
A few minor changes, but overall it's clean and correct.

hashedMae committed:
changes based on notes

hashedMae commented:
changes added 
-maxSupply = 2**128-1, used to ensure _totalSupply doesn't exceed uint128.max, checked during mint() and deposit()
-changed maxMint() and maxDeposit() to reflect new limit on total shares
-changed assets and shares when minting or redeeming to be uint128 instead of uint256
-removed require that prevented minting shares to 0x0
-updated testing patterns based on comments from previous project

hashedMae committed:
typo

hashedMae committed:
correction missed after revert
